sopel/irc.py

```diff
@@ -419,83 +408,54 @@ def found_terminator(self):
 
     def dispatch(self, pretrigger):
         pass
 
-    def error(self, trigger=None):
+    def error(self, trigger=None, exception=None):
         """Called internally when a module causes an error."""
-        try:
-            trace = traceback.format_exc()
-            if sys.version_info.major < 3:
-                trace = trace.decode('utf-8', errors='xmlcharrefreplace')
-            stderr(trace)
-            try:
-                lines = list(reversed(trace.splitlines()))
-                report = [lines[0].strip()]
-                for line in lines:
-                    line = line.strip()
-                    if line.startswith('File "'):
-                        report.append(line[0].lower() + line[1:])
-                        break
-                else:
-                    report.append('source unknown')
-
-                signature = '%s (%s)' % (report[0], report[1])
-                # TODO: make not hardcoded
-                log_filename = os.path.join(self.config.core.logdir, self.config.basename + '.exceptions.log')
-                with codecs.open(log_filename, 'a', encoding='utf-8') as logfile:
-                    logfile.write('Signature: %s\n' % signature)
-                    if trigger:
-                        logfile.write('from {} at {}. Message was: {}\n'.format(
-                            trigger.nick, str(datetime.now()), trigger.group(0)))
-                    logfile.write(trace)
-                    logfile.write(
-                        '----------------------------------------\n\n'
-                    )
-            except Exception as e:
-                stderr("Could not save full traceback!")
-                LOGGER.error("Could not save traceback from %s to file: %s", trigger.sender, str(e))
-
-            if trigger and self.config.core.reply_errors and trigger.sender is not None:
-                self.say(signature, trigger.sender)
-            if trigger:
-                LOGGER.error('Exception from {}: {} ({})'.format(trigger.sender, str(signature), trigger.raw))
-        except Exception as e:
-            if trigger and self.config.core.reply_errors and trigger.sender is not None:
-                self.say("Got an error.", trigger.sender)
-            if trigger:
-                LOGGER.error('Exception from {}: {} ({})'.format(trigger.sender, str(e), trigger.raw))
+        message = 'Unexpected error'
+        if exception:
+            message = '{} ({})'.format(message, exception)
+
+        if trigger:
+            message = '{} from {} at {}. Message was: {}'.format(
+                message, trigger.nick, str(datetime.now()), trigger.group(0)
+            )
+
+        LOGGER.exception(message)
+
+        if trigger and self.config.core.reply_errors and trigger.sender is not None:
+            self.say(message, trigger.sender)
 
     def handle_error(self):
         """Handle any uncaptured error in the core.
 
         Overrides asyncore's handle_error.
```

Comment:
Links to Python docs for `asyncore` and `asyncore.dispatcher.handle_error` would be sugar on top of this. 😁 (Starting point: https://docs.python.org/3/library/asyncore.html#asyncore.dispatcher.handle_error)